packages/std/src/math.rs

```diff
@@ -0,0 +1,255 @@
+use schemars::JsonSchema;
+use serde::{de, ser, Deserialize, Deserializer, Serialize};
+use std::convert::{TryFrom, TryInto};
+use std::{fmt, ops};
+
+use crate::errors::{generic_err, underflow, StdError, StdResult};
+
+/// Decimal9 represents a fixed-point decimal value with 9 fractional digits.
+/// That is Decimal9(1_000_000_000) == 1
+#[derive(Serialize, Deserialize, Copy, Clone, Debug, PartialEq, JsonSchema)]
+pub struct Decimal9(pub u64);
+
+pub const DECIMAL_FRACTIONAL: u64 = 1_000_000_000;
+
+impl Decimal9 {
+    pub fn one() -> Decimal9 {
+        Decimal9(DECIMAL_FRACTIONAL)
+    }
+
+    // convert integer % into Billionth units
+    pub fn percent(percent: u64) -> Decimal9 {
+        Decimal9(percent * 10_000_000)
+    }
+
+    // convert permille (1/1000) into Billionth units
+    pub fn permille(permille: u64) -> Decimal9 {
+        Decimal9(permille * 1_000_000)
+    }
+}
+
+//*** Uint128 ***/
+#[derive(Copy, Clone, Default, Debug, PartialEq, Eq, PartialOrd, Ord, JsonSchema)]
+pub struct Uint128(#[schemars(with = "String")] pub u128);
+
+impl Uint128 {
+    pub fn u128(&self) -> u128 {
+        self.0
+    }
+}
+
+impl From<u128> for Uint128 {
+    fn from(val: u128) -> Self {
+        Uint128(val)
+    }
+}
+
+impl From<u64> for Uint128 {
+    fn from(val: u64) -> Self {
+        Uint128(val.into())
+    }
+}
+
+impl TryFrom<&str> for Uint128 {
+    type Error = StdError;
+
+    fn try_from(val: &str) -> Result<Self, Self::Error> {
+        match val.parse::<u128>() {
+            Ok(u) => Ok(Uint128(u)),
+            Err(e) => Err(generic_err(format!("Parsing coin: {}", e))),
+        }
+    }
+}
+
+impl Into<String> for Uint128 {
+    fn into(self) -> String {
+        self.0.to_string()
+    }
+}
+
+impl Into<u128> for Uint128 {
+    fn into(self) -> u128 {
+        self.u128()
+    }
+}
+
+impl fmt::Display for Uint128 {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl ops::Add for Uint128 {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        Uint128(self.u128() + other.u128())
+    }
+}
+
+impl ops::AddAssign for Uint128 {
+    fn add_assign(&mut self, other: Self) {
+        self.0 += other.u128();
+    }
+}
+
+impl ops::Sub for Uint128 {
+    type Output = StdResult<Self>;
+
+    fn sub(self, other: Self) -> StdResult<Self> {
+        let (min, sub) = (self.u128(), other.u128());
+        if sub > min {
+            Err(underflow(min, sub))
+        } else {
+            Ok(Uint128(min - sub))
+        }
+    }
+}
+
+impl ops::Mul<Decimal9> for Uint128 {
```

Comment:
Where is this used? Would it also be possible to implement `impl ops::Mul<Uint128> for Decimal9` instead, resulting in a Decimal9? Then there would not be any rounding

Reply:
Okay, found it `let tax = amount * invest.exit_tax;`. The file was collapsed.

So if we use `impl ops::Mul<Uint128> for Decimal9`, `tax` becomes a Decimal9 without any rounding issues

Reply:
`tax` needs to be a `Uint128` as we use it to update the balance, which is also a `Uint128`.

We can round down or up. Just define it. In practice, tokens are like `uatom`, so people bond 1000000s of them, so rounding dropping off 0.49 will not make much difference. I added the `min_withdrawal` amount to provide a minimum amount here, to avoid people playing the rounding game.